File: apps/web/src/features/policies/services/workforcePlanService.ts
import { apiClient } from '@/shared/api/api'

/**
 * workforcePlanService — frontend client for B.15b Workforce + Succession (Module 01 § 3).
 */

export type WorkforcePlanStatus = 'draft' | 'active' | 'completed' | 'archived'
export type SuccessionPlanStatus = 'draft' | 'active' | 'archived'
export type Quarter = 'Q1' | 'Q2' | 'Q3' | 'Q4'
export type Criticality = 'alta' | 'media' | 'baja'
export type ReadinessLevel = 1 | 2 | 3 | 4

export interface WorkforcePlan {
  id: string
  tenant?: string | null
  name: string
  fiscal_year: number
  period_start: string
  period_end: string
  description: string
  status: WorkforcePlanStatus
  status_display: string
  owner_user: string
  created_at: string
  updated_at: string
}

export interface HeadcountProjection {
  id: string
  plan: string
  area: string | null
  position: string | null
  current_headcount: number
  projected_headcount: number
  delta_required: number
  target_quarter: Quarter
  justification: string
  created_at: string
  updated_at: string
}

export interface SuccessionPlan {
  id: string
  tenant?: string | null
  name: string
  fiscal_year: number
  notes: string
  status: SuccessionPlanStatus
  status_display: string
  owner_user: string
  created_at: string
  updated_at: string
}

export interface KeyPosition {
  id: string
  plan: string
  position: string
  criticality: Criticality
  criticality_display: string
  risk_notes: string
  current_holder: string | null
  created_at: string
  updated_at: string
}

export interface SuccessorCandidate {
  id: string
  key_position: string
  employee: string
  readiness_level: ReadinessLevel
  readiness_display: string
  order: number
  notes: string
  created_at: string
  updated_at: string
}

interface ApiEnvelope<T> {
  data?: T[] | { results?: T[] } | T
  results?: T[]
}

function unwrapList<T>(resp: { data: ApiEnvelope<T> }): T[] {
  const raw = resp.data
  const fromData = (raw?.data as { results?: T[] } | undefined)?.results
  if (Array.isArray(fromData)) return fromData
  if (Array.isArray(raw?.data)) return raw.data as T[]
  if (Array.isArray(raw?.results)) return raw.results
  return []
}

function unwrapItem<T>(resp: { data: ApiEnvelope<T> }): T {
  const raw = resp.data
  if (
    raw &&
    typeof raw === 'object' &&
    'data' in raw &&
    raw.data &&
    typeof raw.data === 'object' &&
    !Array.isArray(raw.data)
  ) {
    return raw.data as T
  }
  return raw as unknown as T
}

export const workforcePlanService = {
  async listPlans(): Promise<WorkforcePlan[]> {
    const r = await apiClient.get<ApiEnvelope<WorkforcePlan>>(
      '/api/v1/workforce-plans/?page_size=100',
    )
    return unwrapList(r)
  },

  async createPlan(payload: {
    name: string
    fiscal_year: number
    period_start: string
    period_end: string
    owner_user: string
    description?: string
  }): Promise<WorkforcePlan> {
    const r = await apiClient.post<ApiEnvelope<WorkforcePlan>>(
      '/api/v1/workforce-plans/', payload,
    )
    return unwrapItem(r)
  },

  async listProjections(planId?: string): Promise<HeadcountProjection[]> {
    const qs = planId ? `?plan=${planId}&page_size=100` : '?page_size=100'
    const r = await apiClient.get<ApiEnvelope<HeadcountProjection>>(
      `/api/v1/headcount-projections/${qs}`,
    )
    return unwrapList(r)
  },

  async createProjection(payload: {
    plan: string
    current_headcount: number
    projected_headcount: number
    target_quarter: Quarter
    area?: string | null
    position?: string | null
    justification?: string
  }): Promise<HeadcountProjection> {
    const r = await apiClient.post<ApiEnvelope<HeadcountProjection>>(
      '/api/v1/headcount-projections/', payload,
    )
    return unwrapItem(r)
  },

  async listSuccessionPlans(): Promise<SuccessionPlan[]> {
    const r = await apiClient.get<ApiEnvelope<SuccessionPlan>>(
      '/api/v1/succession-plans/?page_size=100',
    )
    return unwrapList(r)
  },

  async createSuccessionPlan(payload: {
    name: string
    fiscal_year: number
    owner_user: string
    notes?: string
  }): Promise<SuccessionPlan> {
    const r = await apiClient.post<ApiEnvelope<SuccessionPlan>>(
      '/api/v1/succession-plans/', payload,
    )
    return unwrapItem(r)
  },

  async listKeyPositions(planId?: string): Promise<KeyPosition[]> {
    const qs = planId ? `?plan=${planId}&page_size=100` : '?page_size=100'
    const r = await apiClient.get<ApiEnvelope<KeyPosition>>(`/api/v1/key-positions/${qs}`)
    return unwrapList(r)
  },

  async createKeyPosition(payload: {
    plan: string
    position: string
    criticality: Criticality
    current_holder?: string | null
    risk_notes?: string
  }): Promise<KeyPosition> {
    const r = await apiClient.post<ApiEnvelope<KeyPosition>>(
      '/api/v1/key-positions/', payload,
    )
    return unwrapItem(r)
  },

  async listCandidates(keyPositionId?: string): Promise<SuccessorCandidate[]> {
    const qs = keyPositionId
      ? `?key_position=${keyPositionId}&page_size=100`
      : '?page_size=100'
    const r = await apiClient.get<ApiEnvelope<SuccessorCandidate>>(
      `/api/v1/successor-candidates/${qs}`,
    )
    return unwrapList(r)
  },

  async createCandidate(payload: {
    key_position: string
    employee: string
    readiness_level: ReadinessLevel
    order?: number
    notes?: string
  }): Promise<SuccessorCandidate> {
    const r = await apiClient.post<ApiEnvelope<SuccessorCandidate>>(
      '/api/v1/successor-candidates/', payload,
    )
    return unwrapItem(r)
  },
}

File: apps/api/apps/policies/services/workforce_plan_service.py
"""Workforce + Succession plan service (B.15b)."""

from __future__ import annotations

from django.core.exceptions import ValidationError

from apps.policies.models import (
    HeadcountProjection,
    KeyPosition,
    SuccessionPlan,
    SuccessorCandidate,
    WorkforcePlan,
)


def create_plan(
    *,
    tenant,
    name: str,
    fiscal_year: int,
    period_start,
    period_end,
    owner_user,
    description: str = '',
) -> WorkforcePlan:
    if period_end < period_start:
        raise ValidationError('period_end debe ser >= period_start.')
    return WorkforcePlan.objects.create(
        tenant=tenant,
        name=name.strip(),
        fiscal_year=fiscal_year,
        period_start=period_start,
        period_end=period_end,
        owner_user=owner_user,
        description=description,
        status='draft',
    )


def add_projection(
    *,
    plan: WorkforcePlan,
    current_headcount: int,
    projected_headcount: int,
    area=None,
    position=None,
    justification: str = '',
    target_quarter: str = 'Q1',
) -> HeadcountProjection:
    if current_headcount < 0 or projected_headcount < 0:
        raise ValidationError('Headcount no puede ser negativo.')
    if target_quarter not in dict(HeadcountProjection.QUARTERS):
        raise ValidationError('target_quarter inválido.')
    return HeadcountProjection.objects.create(
        plan=plan,
        area=area,
        position=position,
        current_headcount=current_headcount,
        projected_headcount=projected_headcount,
        justification=justification,
        target_quarter=target_quarter,
    )


def create_succession_plan(
    *,
    tenant,
    name: str,
    fiscal_year: int,
    owner_user,
    notes: str = '',
) -> SuccessionPlan:
    return SuccessionPlan.objects.create(
        tenant=tenant,
        name=name.strip(),
        fiscal_year=fiscal_year,
        owner_user=owner_user,
        notes=notes,
        status='draft',
    )


def add_key_position(
    *,
    succession_plan: SuccessionPlan,
    position,
    criticality: str = 'media',
    current_holder=None,
    risk_notes: str = '',
) -> KeyPosition:
    if criticality not in dict(KeyPosition.CRITICALITY):
        raise ValidationError('criticality inválida.')
    key_pos, _created = KeyPosition.objects.get_or_create(
        plan=succession_plan,
        position=position,
        defaults={
            'criticality': criticality,
            'current_holder': current_holder,
            'risk_notes': risk_notes,
        },
    )
    return key_pos


def add_successor(
    *,
    key_position: KeyPosition,
    employee,
    readiness_level: int = 4,
    order: int = 0,
    notes: str = '',
) -> SuccessorCandidate:
    if readiness_level not in {1, 2, 3, 4}:
        raise ValidationError('readiness_level debe ser 1..4.')
    candidate, _created = SuccessorCandidate.objects.update_or_create(
        key_position=key_position,
        employee=employee,
        defaults={
            'readiness_level': readiness_level,
            'order': order,
            'notes': notes,
        },
    )
    return candidate
